### ISSUE TITLE
Metrics doesn't work after consistent put or get. [JIRA: RIAK-1575]

### ISSUE THREAD

**Issue by ksauzz** [basho/riak_kv#1086]
`riak-admin status` and `riak-admin stat` fails after consistent put or get. According to crash.log it seems exometer's process died.

crash.log

```
2015-02-27 04:31:44 =CRASH REPORT====
  crasher:
    initial call: exometer_proc:-spawn_process/2-fun-0-/0
    pid: <0.684.0>
    registered_name: []
    exception error: {function_clause,[{exometer_shallowtree,insert,[undefined,166979104625,{t,0,500,[]}],[{file,"src/exometer_shallowtree.erl"},{line,53}]},{exometer_histogram,into_heap,4,[{file,"src/exometer_histogram.erl"},{line,314}]},{exometer_histogram,update_int,3,[{file,"src/exometer_histogram.erl"},{line,309}]},{exometer_histogram,probe_update,2,[{file,"src/exometer_histogram.erl"},{line,303}]},{exometer_probe,handle_msg,2,[{file,"src/exometer_probe.erl"},{line,545}]},{exometer_probe,loop,1,[{file,"src/exometer_probe.erl"},{line,514}]},{proc_lib,init_p,3,[{file,"proc_lib.erl"},{line,224}]}]}
    ancestors: [exometer_admin,exometer_core_sup,<0.154.0>]
    messages: []
    links: []
    dictionary: [{{exometer_proc,info},{info,<0.156.0>,[],{sys,undefined,exometer_probe,[riak,riak_kv,consistent,puts,objsize]}}}]
    trap_exit: false
    status: running
    heap_size: 46422
    stack_size: 27
    reductions: 289
  neighbours:
```
#### To reproduce (riak-2.0.5)

setup strong consistency

```
% sed -i.back 's/## strong_consistency = on/strong_consistency = on/g' riak.config
% riak-admin bucket-type create strongly_consistent '{"props":{"consistent":true}}'
% riak-admin bucket-type activate strongly_consistent
```

put object

```
% curl -XPUT localhost:8098/types/strongly_consistent/buckets/demo/keys/key1 -d 'hi' -i
HTTP/1.1 204 No Content
Vary: Accept-Encoding
Server: MochiWeb/1.1 WebMachine/1.10.5 (jokes are better explained)
Date: Thu, 26 Feb 2015 19:59:23 GMT
Content-Type: application/x-www-form-urlencoded
Content-Length: 0
```

attempt to get metrics

```
% ./riak-admin status
Status failed, see log for details
```

```
% ./riak-admin stat show "**"
RPC to 'riak1@127.0.0.1' failed: {'EXIT',
                                  {noproc,
                                   [{exometer_proc,call,2,
                                     [{file,"src/exometer_proc.erl"},
                                      {line,118}]},
                                    {exometer,get_value,2,
                                     [{file,"src/exometer.erl"},{line,307}]},
                                    {riak_core_console,
                                     '-legacy_search_1/3-fun-1-',4,
                                     [{file,"src/riak_core_console.erl"},
                                      {line,1270}]},
                                    {lists,foldr,3,
                                     [{file,"lists.erl"},{line,1261}]},
                                    {riak_core_console,legacy_search,3,
                                     [{file,"src/riak_core_console.erl"},
                                      {line,1229}]},
                                    {riak_core_console,
                                     '-find_entries/2-fun-0-',2,
                                     [{file,"src/riak_core_console.erl"},
                                      {line,1170}]},
                                    {lists,map,2,
                                     [{file,"lists.erl"},{line,1224}]},
                                    {riak_core_console,stat_show,1,
                                     [{file,"src/riak_core_console.erl"},
                                      {line,1157}]}]}}
```

**Comment by seancribbs** [comment]
@ksauzz Was this fixed in 2.1? Can you re-check?

**Comment by ksauzz** [comment]
@seancribbs I reproduced this on 2.1.

console.log

```
2015-04-17 10:27:24.798 [error] <0.564.0> CRASH REPORT Process <0.564.0> with 0 neighbours crashed with reason: no function clause matching exometer_shallowtree:insert(undefined, 171234044798, {t,0,500,[]}) line 53
2015-04-17 10:27:25.673 [error] <0.13096.0>@riak_kv_console:status:185 Status failed error:noproc
```

crash.log

```
2015-04-17 10:27:24 =CRASH REPORT====
  crasher:
    initial call: exometer_proc:-spawn_process/2-fun-0-/0
    pid: <0.564.0>
    registered_name: []
    exception error: {function_clause,[{exometer_shallowtree,insert,[undefined,171234044798,{t,0,500,[]}],[{file,"src/exometer_shallowtree.erl"},{line,53}]},{exometer_histogram,into_heap,4,[{file,"src/exometer_histogram.erl"},{line,314}]},{exometer_histogram,update_int,3,[{file,"src/exometer_histogram.erl"},{line,309}]},{exometer_histogram,probe_update,2,[{file,"src/exometer_histogram.erl"},{line,303}]},{exometer_probe,handle_msg,2,[{file,"src/exometer_probe.erl"},{line,545}]},{exometer_probe,loop,1,[{file,"src/exometer_probe.erl"},{line,514}]},{proc_lib,init_p,3,[{file,"proc_lib.erl"},{line,224}]}]}
    ancestors: [exometer_admin,exometer_core_sup,<0.163.0>]
    messages: []
    links: []
    dictionary: [{{exometer_proc,info},{info,<0.165.0>,[],{sys,undefined,exometer_probe,[riak,riak_kv,consistent,gets,objsize]}}}]
    trap_exit: false
    status: running
    heap_size: 46422
    stack_size: 27
    reductions: 5213
  neighbours:
```

**Comment by seancribbs** [comment]
Thanks @ksauzz!

**Comment by uwiger** [comment]
The thing that crashes the histogram is that `riak_client:maybe_update_consistent_stat/5` calls `riak_kv_stat:update({Stat, Bucket, Duration, ObjSize})` with `ObjSize=undefined`.

From the looks of it, this is deliberate, but `riak_kv_stat:do_update/1` needs to detect it (as it does with several other metrics), and not try to add `undefined` to the histogram.

There are of course other issues:
- The exometer_histogram implementation needs to reject non-numeric inputs. This has been fixed in a later version of exometer_core
- If there is still some bug in the histogram implementation, and the process dies, it needs to bounce back. This has been fixed in one of the latest versions of exometer_core.

What remains to be fixed, apart from integrating the latest exometer_core, is to add a clause to `riak_kv_stat:do_update/1`, matching on `ObjSize=undefined`.

**Comment by uwiger** [comment]
To be clear:

``` erlang
do_update({consistent_put, _Bucket, Microsecs, ObjSize}) ->
    P = ?PFX,
    ok = exometer:update([P, ?APP, consistent, puts], 1),
    ok = exometer:update([P, ?APP, consistent, puts, time], Microsecs),
    create_or_update([P, ?APP, consistent, puts, objsize], ObjSize, histogram);
```

needs to change to:

``` erlang
do_update({consistent_put, _Bucket, Microsecs, undefined}) ->
    P = ?PFX,
    ok = exometer:update([P, ?APP, consistent, puts], 1),
    ok = exometer:update([P, ?APP, consistent, puts, time], Microsecs);
do_update({consistent_put, _Bucket, Microsecs, ObjSize}) ->
    P = ?PFX,
    ok = exometer:update([P, ?APP, consistent, puts], 1),
    ok = exometer:update([P, ?APP, consistent, puts, time], Microsecs),
    create_or_update([P, ?APP, consistent, puts, objsize], ObjSize, histogram);
```

which is consistent in style with similar patterns above it.

**Comment by uwiger** [comment]
I was able to replicate the crash. Substituting the latest exometer_core for the old one (locally), the crash goes away, even though `riak_kv_stat.erl` hasn't been fixed yet.

The reason for this is that exometer_histogram now discards non-numeric inputs (not even reporting the error). In other words, there is no crash.

We can fake a histogram crash and verify that it recovers:

``` erlang
(dev1@127.0.0.1)1> exometer:info([riak,riak_kv,consistent,puts,objsize],ref).
<0.642.0>
(dev1@127.0.0.1)2> exit(v(1),kill).
true
(dev1@127.0.0.1)3> exometer:info([riak,riak_kv,consistent,puts,objsize],ref).
<0.21780.1>
(dev1@127.0.0.1)4> exometer:get_value([riak,riak_kv,consistent,puts,objsize]).
{ok,[{n,0},
     {mean,0},
     {min,0},
     {max,0},
     {median,0},
     {50,0},
     {75,0},
     {90,0},
     {95,0},
     {99,0},
     {999,0}]}
```

It would still be good form to fix `riak_kv_stat:do_update/1`.

**Comment by kri5** [comment]
Any updates?

**Comment by jonmeredith** [comment]
We've merged the fix into the 2.0 mainline and it'll merge forward.  It should appear in the next 2.0 and 2.1 releases.

**Comment by Basho-JIRA** [comment]
why is this in PMT not Riak

**_[posted via JIRA by Jon Meredith]**_

**Comment by jonmeredith** [comment]
tracked as RIAK-2087

**Comment by Basho-JIRA** [comment]
Via Riak-2087 - We've merged the fix into the 2.0 mainline and it'll merge forward. It should appear in the next 2.0 and 2.1 releases. per Jon M. I am removing the labels on this story Riak 2141 since noted 2087 is tracking but much of the history can be found on 2141, not 2087  

**_[posted via JIRA by Patricia Brewer]**_
